Common subdirectories: tokyocabinet-1.4.34/bros and tokyocabinet-1.4.34-neal/bros
Only in tokyocabinet-1.4.34-neal: config.log
Only in tokyocabinet-1.4.34-neal: config.status
Common subdirectories: tokyocabinet-1.4.34/doc and tokyocabinet-1.4.34-neal/doc
Common subdirectories: tokyocabinet-1.4.34/example and tokyocabinet-1.4.34-neal/example
Common subdirectories: tokyocabinet-1.4.34/lab and tokyocabinet-1.4.34-neal/lab
Only in tokyocabinet-1.4.34-neal: libtokyocabinet.a
Only in tokyocabinet-1.4.34-neal: libtokyocabinet.so
Only in tokyocabinet-1.4.34-neal: libtokyocabinet.so.8
Only in tokyocabinet-1.4.34-neal: libtokyocabinet.so.8.25.0
Only in tokyocabinet-1.4.34-neal: Makefile
Only in tokyocabinet-1.4.34-neal: Makefile.TRP
Common subdirectories: tokyocabinet-1.4.34/man and tokyocabinet-1.4.34-neal/man
Only in tokyocabinet-1.4.34-neal: md5.o
Only in tokyocabinet-1.4.34-neal: myconf.o
Only in tokyocabinet-1.4.34-neal: .svn
diff -up tokyocabinet-1.4.34/tcadb.c tokyocabinet-1.4.34-neal/tcadb.c
--- tokyocabinet-1.4.34/tcadb.c	2009-10-01 05:28:57.000000000 -0600
+++ tokyocabinet-1.4.34-neal/tcadb.c	2009-10-07 16:21:49.000000000 -0600
@@ -114,6 +114,7 @@ bool tcadbopen(TCADB *adb, const char *n
   int32_t nmemb = -1;
   int32_t lcnum = -1;
   int32_t ncnum = -1;
+  int32_t rssgoal = -1;
   int32_t width = -1;
   int64_t limsiz = -1;
   TCLIST *idxs = NULL;
@@ -158,6 +159,8 @@ bool tcadbopen(TCADB *adb, const char *n
       nmemb = tcatoix(pv);
     } else if(!tcstricmp(elem, "lcnum")){
       lcnum = tcatoix(pv);
+    } else if(!tcstricmp(elem, "rssgoal")){
+      rssgoal = tcatoix(pv);
     } else if(!tcstricmp(elem, "ncnum")){
       ncnum = tcatoix(pv);
     } else if(!tcstricmp(elem, "width")){
@@ -228,7 +231,7 @@ bool tcadbopen(TCADB *adb, const char *n
     if(tbmode) opts |= BDBTBZIP;
     if(ttmode) opts |= BDBTTCBS;
     tcbdbtune(bdb, lmemb, nmemb, bnum, apow, fpow, opts);
-    tcbdbsetcache(bdb, lcnum, ncnum);
+    tcbdbsetcache(bdb, lcnum, ncnum, rssgoal);
     if(xmsiz >= 0) tcbdbsetxmsiz(bdb, xmsiz);
     if(dfunit >= 0) tcbdbsetdfunit(bdb, dfunit);
     if(capnum > 0) tcbdbsetcapnum(bdb, capnum);
Only in tokyocabinet-1.4.34-neal: tcadb.o
Only in tokyocabinet-1.4.34-neal: tcamgr
Only in tokyocabinet-1.4.34-neal: tcamgr.o
Only in tokyocabinet-1.4.34-neal: tcatest
Only in tokyocabinet-1.4.34-neal: tcatest.o
Only in tokyocabinet-1.4.34-neal: tcawmgr.cgi
Only in tokyocabinet-1.4.34-neal: tcawmgr.o
diff -up tokyocabinet-1.4.34/tcbdb.c tokyocabinet-1.4.34-neal/tcbdb.c
--- tokyocabinet-1.4.34/tcbdb.c	2009-10-01 05:28:57.000000000 -0600
+++ tokyocabinet-1.4.34-neal/tcbdb.c	2009-10-12 09:09:38.000000000 -0600
@@ -13,6 +13,7 @@
  * Boston, MA 02111-1307 USA.
  *************************************************************************************************/
 
+#include <unistd.h>
 
 #include "tcutil.h"
 #include "tchdb.h"
@@ -25,6 +26,8 @@
 #define BDBNODEIDBASE  ((1LL<<48)+1)     // base number of node ID
 #define BDBLEVELMAX    64                // max level of B+ tree
 #define BDBCACHEOUT    8                 // number of pages in a process of cacheout
+#define RSSUNLIMITED   INT_MAX           // rss management - unlimited memory
+#define RSSMINIMUM     8388608           // rss management - minimum memory size 8MB
 
 #define BDBDEFLMEMB    128               // default number of members in each leaf
 #define BDBMINLMEMB    4                 // minimum number of members in each leaf
@@ -270,7 +273,7 @@ bool tcbdbtune(TCBDB *bdb, int32_t lmemb
 
 
 /* Set the caching parameters of a B+ tree database object. */
-bool tcbdbsetcache(TCBDB *bdb, int32_t lcnum, int32_t ncnum){
+bool tcbdbsetcache(TCBDB *bdb, int32_t lcnum, int32_t ncnum, int32_t rssgoal){
   assert(bdb);
   if(bdb->open){
     tcbdbsetecode(bdb, TCEINVALID, __FILE__, __LINE__, __func__);
@@ -278,6 +281,8 @@ bool tcbdbsetcache(TCBDB *bdb, int32_t l
   }
   if(lcnum > 0) bdb->lcnum = tclmax(lcnum, BDBLEVELMAX);
   if(ncnum > 0) bdb->ncnum = tclmax(ncnum, BDBLEVELMAX);
+  if(rssgoal > 0) bdb->rssgoal = tclmax(tclmin(rssgoal, RSSUNLIMITED), RSSMINIMUM);
+
   return true;
 }
 
@@ -497,7 +502,7 @@ void *tcbdbget(TCBDB *bdb, const void *k
   } else {
     rv = NULL;
   }
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     if(!bdb->tran && !tcbdbcacheadjust(bdb)){
@@ -528,7 +533,7 @@ const void *tcbdbget3(TCBDB *bdb, const 
     return NULL;
   }
   const char *rv = tcbdbgetimpl(bdb, kbuf, ksiz, sp);
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     if(!bdb->tran && !tcbdbcacheadjust(bdb)) rv = NULL;
@@ -548,7 +553,7 @@ TCLIST *tcbdbget4(TCBDB *bdb, const void
     return NULL;
   }
   TCLIST *rv = tcbdbgetlist(bdb, kbuf, ksiz);
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     if(!bdb->tran && !tcbdbcacheadjust(bdb)){
@@ -571,7 +576,7 @@ int tcbdbvnum(TCBDB *bdb, const void *kb
     return 0;
   }
   int rv = tcbdbgetnum(bdb, kbuf, ksiz);
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     if(!bdb->tran && !tcbdbcacheadjust(bdb)) rv = 0;
@@ -616,7 +621,7 @@ TCLIST *tcbdbrange(TCBDB *bdb, const voi
     return keys;
   }
   tcbdbrangeimpl(bdb, bkbuf, bksiz, binc, ekbuf, eksiz, einc, max, keys);
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     tcbdbcacheadjust(bdb);
@@ -646,7 +651,7 @@ TCLIST *tcbdbfwmkeys(TCBDB *bdb, const v
     return keys;
   }
   tcbdbrangefwm(bdb, pbuf, psiz, max, keys);
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     tcbdbcacheadjust(bdb);
@@ -973,7 +978,7 @@ bool tcbdbcurfirst(BDBCUR *cur){
     return false;
   }
   bool rv = tcbdbcurfirstimpl(cur);
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     if(!bdb->tran && !tcbdbcacheadjust(bdb)) rv = false;
@@ -994,7 +999,7 @@ bool tcbdbcurlast(BDBCUR *cur){
     return false;
   }
   bool rv = tcbdbcurlastimpl(cur);
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     if(!bdb->tran && !tcbdbcacheadjust(bdb)) rv = false;
@@ -1015,7 +1020,7 @@ bool tcbdbcurjump(BDBCUR *cur, const voi
     return false;
   }
   bool rv = tcbdbcurjumpimpl(cur, kbuf, ksiz, true);
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     if(!bdb->tran && !tcbdbcacheadjust(bdb)) rv = false;
@@ -1048,7 +1053,7 @@ bool tcbdbcurprev(BDBCUR *cur){
     return false;
   }
   bool rv = tcbdbcurprevimpl(cur);
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     if(!bdb->tran && !tcbdbcacheadjust(bdb)) rv = false;
@@ -1074,7 +1079,7 @@ bool tcbdbcurnext(BDBCUR *cur){
     return false;
   }
   bool rv = tcbdbcurnextimpl(cur);
-  bool adj = TCMAPRNUM(bdb->leafc) > bdb->lcnum || TCMAPRNUM(bdb->nodec) > bdb->ncnum;
+  bool adj = TCLEAFC_OVERBUDGET(bdb) || TCNODEC_OVERBUDGET(bdb) || TCRSS_OVERBUDGET(bdb);
   BDBUNLOCKMETHOD(bdb);
   if(adj && BDBLOCKMETHOD(bdb, true)){
     if(!bdb->tran && !tcbdbcacheadjust(bdb)) rv = false;
@@ -2859,11 +2864,26 @@ static bool tcbdbcacheadjust(TCBDB *bdb)
     bool clk = BDBLOCKCACHE(bdb);
     TCMAP *leafc = bdb->leafc;
     tcmapiterinit(leafc);
-    int dnum = tclmax(TCMAPRNUM(bdb->leafc) - bdb->lcnum, BDBCACHEOUT);
+
+    uint32_t tmp_leafc = TCMAPRNUM(bdb->leafc);
+    uint32_t rssnow = tcgetprocrss();
+    uint32_t lcdiff = tmp_leafc - bdb->lcnum;
+
+    float rssgoalratio = (rssnow/bdb->rssgoal);
+    uint32_t rsslcdiff = 0;
+
+    if((bdb->rssgoal > 1) && ( rssgoalratio > 1.10))
+        rsslcdiff = (int)((rssgoalratio - 1) * tmp_leafc);
+
+    int dnum = tclmax(tclmax(lcdiff, rsslcdiff), BDBCACHEOUT);
+
     for(int i = 0; i < dnum; i++){
       int rsiz;
       if(!tcbdbleafcacheout(bdb, (BDBLEAF *)tcmapiterval(tcmapiternext(leafc, &rsiz), &rsiz)))
         err = true;
+
+    //fprintf(stderr, "tcbdbcacheadjust (leafc) oldrss=[%ld], newrss[%ld], rssgoal=[%ld]\n", rssnow, tcgetprocrss(), bdb->rssgoal);
+
     }
     if(clk) BDBUNLOCKCACHE(bdb);
     if(!err && tchdbecode(bdb->hdb) != ecode)
@@ -2875,7 +2895,20 @@ static bool tcbdbcacheadjust(TCBDB *bdb)
     bool clk = BDBLOCKCACHE(bdb);
     TCMAP *nodec = bdb->nodec;
     tcmapiterinit(nodec);
-    int dnum = tclmax(TCMAPRNUM(bdb->nodec) - bdb->ncnum, BDBCACHEOUT);
+
+
+    uint32_t tmp_nodec = TCMAPRNUM(bdb->nodec);
+    uint32_t rssnow = tcgetprocrss();
+    uint32_t ncdiff = tmp_nodec - bdb->ncnum;
+
+    float rssgoalratio = (rssnow/bdb->rssgoal);
+    uint32_t rssncdiff = 0;
+
+    if((bdb->rssgoal > 1) && ( rssgoalratio > 1.10))
+        rssncdiff = (int)((rssgoalratio - 1) * tmp_nodec);
+
+    int dnum = tclmax(tclmax(ncdiff, rssncdiff), BDBCACHEOUT);
+
     for(int i = 0; i < dnum; i++){
       int rsiz;
       if(!tcbdbnodecacheout(bdb, (BDBNODE *)tcmapiterval(tcmapiternext(nodec, &rsiz), &rsiz)))
@@ -2884,7 +2917,11 @@ static bool tcbdbcacheadjust(TCBDB *bdb)
     if(clk) BDBUNLOCKCACHE(bdb);
     if(!err && tchdbecode(bdb->hdb) != ecode)
       tcbdbsetecode(bdb, ecode, __FILE__, __LINE__, __func__);
+
+    //fprintf(stderr, "tcbdbcacheadjust (nodec) oldrss=[%ld], newrss[%ld], rssgoal=[%ld]\n", rssnow, tcgetprocrss(), bdb->rssgoal);
   }
+
+
   return !err;
 }
 
@@ -3438,7 +3475,7 @@ static bool tcbdboptimizeimpl(TCBDB *bdb
   if(fpow < 0) fpow = tclog2l(tchdbfbpmax(bdb->hdb));
   if(opts == UINT8_MAX) opts = bdb->opts;
   tcbdbtune(tbdb, lmemb, nmemb, bnum, apow, fpow, opts);
-  tcbdbsetcache(tbdb, 1, 1);
+  tcbdbsetcache(tbdb, 1, 1, 1);
   tcbdbsetlsmax(tbdb, bdb->lsmax);
   uint32_t lcnum = bdb->lcnum;
   uint32_t ncnum = bdb->ncnum;
@@ -4079,6 +4116,7 @@ void tcbdbprintmeta(TCBDB *bdb){
   wp += sprintf(wp, " cmpop=%p", (void *)bdb->cmpop);
   wp += sprintf(wp, " lcnum=%u", bdb->lcnum);
   wp += sprintf(wp, " ncnum=%u", bdb->ncnum);
+  wp += sprintf(wp, " rssgoal=%u", bdb->rssgoal);
   wp += sprintf(wp, " lsmax=%u", bdb->lsmax);
   wp += sprintf(wp, " lschk=%u", bdb->lschk);
   wp += sprintf(wp, " capnum=%llu", (unsigned long long)bdb->capnum);
diff -up tokyocabinet-1.4.34/tcbdb.h tokyocabinet-1.4.34-neal/tcbdb.h
--- tokyocabinet-1.4.34/tcbdb.h	2009-10-01 05:28:57.000000000 -0600
+++ tokyocabinet-1.4.34-neal/tcbdb.h	2009-10-07 16:21:49.000000000 -0600
@@ -59,6 +59,7 @@ typedef struct {                        
   void *cmpop;                           /* opaque object for the comparison function */
   uint32_t lcnum;                        /* maximum number of cached leaves */
   uint32_t ncnum;                        /* maximum number of cached nodes */
+  uint32_t rssgoal;                      /* desired rss size */
   uint32_t lsmax;                        /* maximum size of each leaf */
   uint32_t lschk;                        /* counter for leaf size checking */
   uint64_t capnum;                       /* capacity number of records */
@@ -206,7 +207,7 @@ bool tcbdbtune(TCBDB *bdb, int32_t lmemb
    the default value is specified.  The default value is 512.
    If successful, the return value is true, else, it is false.
    Note that the caching parameters should be set before the database is opened. */
-bool tcbdbsetcache(TCBDB *bdb, int32_t lcnum, int32_t ncnum);
+bool tcbdbsetcache(TCBDB *bdb, int32_t lcnum, int32_t ncnum, int32_t rssgoal);
 
 
 /* Set the size of the extra mapped memory of a B+ tree database object.
Only in tokyocabinet-1.4.34-neal: tcbdb.o
Only in tokyocabinet-1.4.34-neal: tcbmgr
Only in tokyocabinet-1.4.34-neal: tcbmgr.o
Only in tokyocabinet-1.4.34-neal: tcbmttest
Only in tokyocabinet-1.4.34-neal: tcbmttest.o
Only in tokyocabinet-1.4.34-neal: tcbtest
diff -up tokyocabinet-1.4.34/tcbtest.c tokyocabinet-1.4.34-neal/tcbtest.c
--- tokyocabinet-1.4.34/tcbtest.c	2009-10-01 05:28:57.000000000 -0600
+++ tokyocabinet-1.4.34-neal/tcbtest.c	2009-10-07 16:21:49.000000000 -0600
@@ -790,7 +790,7 @@ static int procwrite(const char *path, i
     eprint(bdb, __LINE__, "tcbdbtune");
     err = true;
   }
-  if(!tcbdbsetcache(bdb, lcnum, ncnum)){
+  if(!tcbdbsetcache(bdb, lcnum, ncnum,0)){
     eprint(bdb, __LINE__, "tcbdbsetcache");
     err = true;
   }
@@ -879,7 +879,7 @@ static int procread(const char *path, bo
     eprint(bdb, __LINE__, "tcbdbsetcodecfunc");
     err = true;
   }
-  if(!tcbdbsetcache(bdb, lcnum, ncnum)){
+  if(!tcbdbsetcache(bdb, lcnum, ncnum,0)){
     eprint(bdb, __LINE__, "tcbdbsetcache");
     err = true;
   }
@@ -972,7 +972,7 @@ static int procremove(const char *path, 
     eprint(bdb, __LINE__, "tcbdbsetcodecfunc");
     err = true;
   }
-  if(!tcbdbsetcache(bdb, lcnum, ncnum)){
+  if(!tcbdbsetcache(bdb, lcnum, ncnum,0)){
     eprint(bdb, __LINE__, "tcbdbsetcache");
     err = true;
   }
@@ -1063,7 +1063,7 @@ static int procrcat(const char *path, in
     eprint(bdb, __LINE__, "tcbdbtune");
     err = true;
   }
-  if(!tcbdbsetcache(bdb, lcnum, ncnum)){
+  if(!tcbdbsetcache(bdb, lcnum, ncnum,0)){
     eprint(bdb, __LINE__, "tcbdbsetcache");
     err = true;
   }
@@ -1256,7 +1256,7 @@ static int procqueue(const char *path, i
     eprint(bdb, __LINE__, "tcbdbtune");
     err = true;
   }
-  if(!tcbdbsetcache(bdb, lcnum, ncnum)){
+  if(!tcbdbsetcache(bdb, lcnum, ncnum,0)){
     eprint(bdb, __LINE__, "tcbdbsetcache");
     err = true;
   }
@@ -1376,7 +1376,7 @@ static int procmisc(const char *path, in
     eprint(bdb, __LINE__, "tcbdbtune");
     err = true;
   }
-  if(!tcbdbsetcache(bdb, 128, 256)){
+  if(!tcbdbsetcache(bdb, 128, 256, 0)){
     eprint(bdb, __LINE__, "tcbdbsetcache");
     err = true;
   }
@@ -2203,7 +2203,7 @@ static int procwicked(const char *path, 
     eprint(bdb, __LINE__, "tcbdbtune");
     err = true;
   }
-  if(!tcbdbsetcache(bdb, 128, 256)){
+  if(!tcbdbsetcache(bdb, 128, 256, 0)){
     eprint(bdb, __LINE__, "tcbdbsetcache");
     err = true;
   }
Only in tokyocabinet-1.4.34-neal: tcbtest.o
Only in tokyocabinet-1.4.34-neal: tcfdb.o
Only in tokyocabinet-1.4.34-neal: tcfmgr
Only in tokyocabinet-1.4.34-neal: tcfmgr.o
Only in tokyocabinet-1.4.34-neal: tcfmttest
Only in tokyocabinet-1.4.34-neal: tcfmttest.o
Only in tokyocabinet-1.4.34-neal: tcftest
Only in tokyocabinet-1.4.34-neal: tcftest.o
Only in tokyocabinet-1.4.34-neal: tchdb.o
Only in tokyocabinet-1.4.34-neal: tchmgr
Only in tokyocabinet-1.4.34-neal: tchmgr.o
Only in tokyocabinet-1.4.34-neal: tchmttest
Only in tokyocabinet-1.4.34-neal: tchmttest.o
Only in tokyocabinet-1.4.34-neal: tchtest
Only in tokyocabinet-1.4.34-neal: tchtest.o
diff -up tokyocabinet-1.4.34/tctdb.c tokyocabinet-1.4.34-neal/tctdb.c
--- tokyocabinet-1.4.34/tctdb.c	2009-10-01 05:28:57.000000000 -0600
+++ tokyocabinet-1.4.34-neal/tctdb.c	2009-10-07 16:21:49.000000000 -0600
@@ -1989,7 +1989,7 @@ static bool tctdbopenimpl(TCTDB *tdb, co
       if(dbgfd >= 0) tcbdbsetdbgfd(bdb, dbgfd);
       if(tdb->mmtx) tcbdbsetmutex(bdb);
       if(enc && dec) tcbdbsetcodecfunc(bdb, enc, encop, dec, decop);
-      tcbdbsetcache(bdb, tdb->lcnum, tdb->ncnum);
+      tcbdbsetcache(bdb, tdb->lcnum, tdb->ncnum, 0);
       tcbdbsetxmsiz(bdb, tchdbxmsiz(tdb->hdb));
       tcbdbsetdfunit(bdb, tchdbdfunit(tdb->hdb));
       tcbdbsetlsmax(bdb, TDBIDXLSMAX);
@@ -2679,7 +2679,7 @@ static bool tctdbsetindeximpl(TCTDB *tdb
     if(tdb->mmtx) tcbdbsetmutex(idx->db);
     if(enc && dec) tcbdbsetcodecfunc(idx->db, enc, encop, dec, decop);
     tcbdbtune(idx->db, TDBIDXLMEMB, TDBIDXNMEMB, bbnum, -1, -1, bopts);
-    tcbdbsetcache(idx->db, tdb->lcnum, tdb->ncnum);
+    tcbdbsetcache(idx->db, tdb->lcnum, tdb->ncnum, 0);
     tcbdbsetxmsiz(idx->db, bxmsiz);
     tcbdbsetdfunit(idx->db, tchdbdfunit(tdb->hdb));
     tcbdbsetlsmax(idx->db, TDBIDXLSMAX);
@@ -2698,7 +2698,7 @@ static bool tctdbsetindeximpl(TCTDB *tdb
     tcbdbsetcmpfunc(idx->db, tccmpdecimal, NULL);
     if(enc && dec) tcbdbsetcodecfunc(idx->db, enc, encop, dec, decop);
     tcbdbtune(idx->db, TDBIDXLMEMB, TDBIDXNMEMB, bbnum, -1, -1, bopts);
-    tcbdbsetcache(idx->db, tdb->lcnum, tdb->ncnum);
+    tcbdbsetcache(idx->db, tdb->lcnum, tdb->ncnum, 0);
     tcbdbsetxmsiz(idx->db, bxmsiz);
     tcbdbsetdfunit(idx->db, tchdbdfunit(tdb->hdb));
     tcbdbsetlsmax(idx->db, TDBIDXLSMAX);
@@ -2717,7 +2717,7 @@ static bool tctdbsetindeximpl(TCTDB *tdb
     if(tdb->mmtx) tcbdbsetmutex(idx->db);
     if(enc && dec) tcbdbsetcodecfunc(idx->db, enc, encop, dec, decop);
     tcbdbtune(idx->db, TDBIDXLMEMB, TDBIDXNMEMB, bbnum, -1, -1, bopts);
-    tcbdbsetcache(idx->db, tdb->lcnum, tdb->ncnum);
+    tcbdbsetcache(idx->db, tdb->lcnum, tdb->ncnum, 0);
     tcbdbsetxmsiz(idx->db, bxmsiz);
     tcbdbsetdfunit(idx->db, tchdbdfunit(tdb->hdb));
     tcbdbsetlsmax(idx->db, TDBIDXLSMAX);
@@ -2736,7 +2736,7 @@ static bool tctdbsetindeximpl(TCTDB *tdb
     if(tdb->mmtx) tcbdbsetmutex(idx->db);
     if(enc && dec) tcbdbsetcodecfunc(idx->db, enc, encop, dec, decop);
     tcbdbtune(idx->db, TDBIDXLMEMB, TDBIDXNMEMB, bbnum, -1, -1, bopts);
-    tcbdbsetcache(idx->db, tdb->lcnum, tdb->ncnum);
+    tcbdbsetcache(idx->db, tdb->lcnum, tdb->ncnum, 0);
     tcbdbsetxmsiz(idx->db, bxmsiz);
     tcbdbsetdfunit(idx->db, tchdbdfunit(tdb->hdb));
     tcbdbsetlsmax(idx->db, TDBIDXLSMAX);
Only in tokyocabinet-1.4.34-neal: tctdb.o
Only in tokyocabinet-1.4.34-neal: tctmgr
Only in tokyocabinet-1.4.34-neal: tctmgr.o
Only in tokyocabinet-1.4.34-neal: tctmttest
Only in tokyocabinet-1.4.34-neal: tctmttest.o
Only in tokyocabinet-1.4.34-neal: tcttest
Only in tokyocabinet-1.4.34-neal: tcttest.o
Only in tokyocabinet-1.4.34-neal: tcucodec
Only in tokyocabinet-1.4.34-neal: tcucodec.o
Only in tokyocabinet-1.4.34-neal: tcumttest
Only in tokyocabinet-1.4.34-neal: tcumttest.o
Only in tokyocabinet-1.4.34-neal: tcutest
Only in tokyocabinet-1.4.34-neal: tcutest.o
diff -up tokyocabinet-1.4.34/tcutil.c tokyocabinet-1.4.34-neal/tcutil.c
--- tokyocabinet-1.4.34/tcutil.c	2009-10-01 05:28:57.000000000 -0600
+++ tokyocabinet-1.4.34-neal/tcutil.c	2009-10-07 16:21:49.000000000 -0600
@@ -13,7 +13,7 @@
  * Boston, MA 02111-1307 USA.
  *************************************************************************************************/
 
-
+#include <proc/readproc.h>
 #include "tcutil.h"
 #include "myconf.h"
 #include "md5.h"
@@ -88,6 +88,36 @@ void tcfree(void *ptr){
   TCFREE(ptr);
 }
 
+/* NEAL:  get RSS process memory size */
+uint32_t tcgetprocrss()
+{
+  pid_t my_pid = -1;
+  proc_t my_proc;
+  my_pid = getpgrp();
+  get_proc_stats(my_pid, &my_proc);
+  
+  return my_proc.rss;
+}
+
+bool tcrssoverbudget(uint32_t rssgoal)
+{
+  pid_t my_pid = -1;
+  proc_t my_proc;
+  my_pid = getpgrp();
+  get_proc_stats(my_pid, &my_proc);
+
+  if((my_proc.rss > 0) && (rssgoal > 1))
+  {
+     float rssgoalratio = (my_proc.rss/rssgoal);
+     if(( rssgoalratio > 1.10))
+        return true;
+  }
+
+  return false;
+}
+
+
+
 
 
 /*************************************************************************************************
diff -up tokyocabinet-1.4.34/tcutil.h tokyocabinet-1.4.34-neal/tcutil.h
--- tokyocabinet-1.4.34/tcutil.h	2009-10-01 05:28:57.000000000 -0600
+++ tokyocabinet-1.4.34-neal/tcutil.h	2009-10-10 12:23:07.000000000 -0600
@@ -109,6 +109,13 @@ char *tcstrdup(const void *str);
 void tcfree(void *ptr);
 
 
+/*************************************************************************************************
+ * Neal's changes
+ *************************************************************************************************/
+
+/* NEAL:  get RSS process memory size */
+uint32_t tcgetprocrss();
+bool tcrssoverbudget(uint32_t rssgoal);
 
 /*************************************************************************************************
  * basic utilities (for experts)
@@ -3880,6 +3887,9 @@ uint64_t tcpagealign(uint64_t off);
   } while(false);
 #endif
 
+#define  TCNODEC_OVERBUDGET(DBP) (((DBP->nodec)->rnum) > DBP->ncnum)
+#define  TCLEAFC_OVERBUDGET(DBP) (((DBP->leafc)->rnum) > DBP->lcnum)
+#define  TCRSS_OVERBUDGET(DBP) tcrssoverbudget(DBP->rssgoal)
 
 /* Print hexadecimal pattern of a binary region. */
 #define TCPRINTHEX(TC_ptr, TC_size) \
Only in tokyocabinet-1.4.34-neal: tcutil.o
Only in tokyocabinet-1.4.34-neal: tokyocabinet.pc
